portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,44 +45,135 @@
+    String parsedBody = getBody(request);
+    String trimmedBody = parsedBody.substring(1, parsedBody.length() - 1);
+    Map<String, String> immutableMap = Splitter.on('&').trimResults().withKeyValueSeparator('=').split(trimmedBody);
+    HashMap<String, String> mutableMap = new HashMap<>();
+    for (String key : immutableMap.keySet()) {
+      String decodedVal = java.net.URLDecoder.decode(immutableMap.get(key), StandardCharsets.UTF_8.name());
+      mutableMap.put(key, decodedVal);
+    }
+    String userComment = getFieldFromMap(mutableMap, "comment", "");
     if(userComment.length() != 0) {
-      String userName = getFieldFromResponse(request, "name", "Anonymous");
-      String userEmail = getFieldFromResponse(request, "email", "janedoe@gmail.com");
+      String userName = getFieldFromMap(mutableMap, "name", "Anonymous");
+      String userEmail = getFieldFromMap(mutableMap, "email", "janedoe@gmail.com");
       Date date = new Date();
-      String currDate = date.toString();
-      String userDate = getFieldFromResponse(request, "time", currDate);
+      TimeZone tz = TimeZone.getTimeZone("UTC");
+      DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'"); 
+      df.setTimeZone(tz);
+      String currDate = df.format(date).toString();
+      String userDate = getFieldFromMap(mutableMap, "timestamp", currDate);
+      addToDatastore(userName, userEmail, userDate, userComment);
       UserComment comment = UserComment.create(userName, userEmail, userComment, userDate);
       userComments.add(comment);
     }
-    response.sendRedirect("index.html");
   }
 
   /*
-   * Extracts the value of fieldName attribute from request object if present
-   * and returns defaultValue if it is not or it is empty
+   * Extracts the value of fieldName attribute from valueMap if present
+   * and returns defaultValue if it is not or the value is empty
    */
-    private String getFieldFromResponse(HttpServletRequest request, String fieldName, String defaultValue) {
+  private String getFieldFromMap(Map<String, String> valueMap, String fieldName, String defaultValue) {
+    String fieldValue = valueMap.getOrDefault(fieldName, defaultValue);
+    if(fieldValue.length() == 0)
+    {
+        fieldValue = defaultValue;
+    }
+    return fieldValue;
+  }
+
+  // Adds a comment with the given metadata to the database  
+  private void addToDatastore(String name, String email, String dateTime, String comment) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("name", name);
+    commentEntity.setProperty("email", email);
+    commentEntity.setProperty("time", dateTime);
+    commentEntity.setProperty("comment", comment);
+    datastore.put(commentEntity);
+  }
+
+  // Parses body of POST request into string
+  public static String getBody(HttpServletRequest request) throws IOException {
```

Comment:
Guava has some helpers for this:

```
CharStreams.toString(request.getReader());
```

https://guava.dev/releases/19.0/api/docs/com/google/common/io/CharStreams.html#toString(java.lang.Readable)